### DISCUSSION TITLE
Different handlers for different bots?

### DISCUSSION THREAD

**Discussion by intoeetive** [def-studio/telegraph#89] [Ideas]
Hi @fabio-ivona 

First of all - this is an excellent piece of software! I am writing telegram bot for a volunteer project, and it's really speeding up the development!

Is it possible use different Webhook Handlers for different bots? So that in config file instead of
`'webhook_handler' => MyWebhookHandler::class,`
I could have something like:
```
'webhook_handlers' => [
    'user_bot' => MyWebhookHandler::class,
    'admin_bot' => MyAdminWebhookHandler::class,
```

I think this might be a lot more convenient then figuring out where the request came from in handler itself

Also been wondering if you have plans to support `request_contact` button anytime soon? (if not, I might submit a PR for that in next few days)

**Comment by fabio-ivona** [comment]
Hi! @intoeetive !

thanks for your kind words, we are glad you find Telegraph useful

Dedicated webhook handlers are not currently supported, but you could achieve something similar by defining a main webhook handler like this one: 

```php

class MainWebhookHandler extends WebhookHandler
{
   public function handle(Request $request, TelegraphBot $bot): void
   {
        match($bot->name){
            'user_bot' => app(MyWebhookHandler::class)->handle($request, $bot),
            'admin_bot' => app(MyAdminWebhookHandler::class)->handle($request, $bot),
            default => throw new Exception('unsupported bot');
       }
   }
}
```

**Reply by intoeetive** [reply]
Oh that's handy, did not think of that approach. Thank you!

**Reply by ah-rahimi** [reply]
@fabio-ivona 
Excuse me
In MyAdminWebhookHandler class & MyWebhookHandler class we must extends WebHookHandler ?

**Reply by fabio-ivona** [reply]
Yeah, so it will inherit all its methods

**Comment by fabio-ivona** [comment]
> Also been wondering if you have plans to support request_contact button anytime soon? (if not, I might submit a PR for that in next few days)

we currently support only inline keyboards, but I am working on `reply keyboards` and they will be released in a few days. As soon as it is ready your PR for request_contact is welcome! thanks!

**Reply by intoeetive** [reply]
That is awesome to hear! Looking forward to it, as it's really needed (my use case it that people need to fill in short questionary before they can use the features of the bot, and reply keyboards including "share phone number" is really key feature for that)

**Reply by fabio-ivona** [reply]
will take our best efforts to get this done as soon as possible

as the package is widely used, we should be care that everything works before releasing it

**Reply by intoeetive** [reply]
@fabio-ivona Hi Fabio - do you have any update on timeline of when `reply keyboards` will be available? (maybe as beta or even alpha version)?

**Reply by fabio-ivona** [reply]
Hi @intoeetive I think we will merge it on master branch in a couple of days (so you can get it by requiring dev-master as beta version) and will release a stable version by the end of the week

**Reply by intoeetive** [reply]
@fabio-ivona i don't see that branch on GitHub, is there any chance it's not pushed?

**Reply by fabio-ivona** [reply]
Not yet pushed, we are refining it and it will be available on a couple of days. Will post an update as soon as it is available

**Reply by fabio-ivona** [reply]
@intoeetive https://github.com/def-studio/telegraph/pull/94 :wink: 

you can require this version for an early test: 

```
composer require defstudio/telegraph:dev-reply-keyboards
```

let me know if it works!

**Reply by intoeetive** [reply]
Thank you - pulling it in works, I'll give actual test today or tomorrow
